torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1598,55 +1600,39 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::pointToPoint(
   return work;
 }
 
-std::vector<at::Tensor> copyCuda(const std::vector<at::Tensor>& inputs) {
+std::vector<at::Tensor> extractCudaTensors(const std::vector<at::Tensor>& inputs) {
   std::vector<at::Tensor> cudaInputs;
   cudaInputs.reserve(inputs.size());
   for (auto& input : inputs) {
-    cudaInputs.push_back(input.to(c10::Device(c10::kCUDA, input.device().index())));
+    cudaInputs.push_back(
+        dynamic_cast<torch_lazy_tensors::compiler::TSData*>(torch::lazy::GetLtcTensor(input).GetDataHandle().get())->data());
   }
 
   return cudaInputs;
 }
 
-void copyLazy(const std::vector<at::Tensor>& lazyInputs, const std::vector<at::Tensor>& cudaInputs) {
-  for (ssize_t i = 0; i < lazyInputs.size(); i++) {
-    lazyInputs[i].copy_(cudaInputs[i]);
-  }
-}
-
 template <typename Fn>
 c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
     std::vector<at::Tensor>& inputs,
     std::vector<at::Tensor>& outputs,
     Fn fn,
     OpType opType,
     const char* profilingTitle) {
-  // A hack to copy lazy to cuda to bypass following checks.
+  // A hack to extract cuda tensors from the lazy tensor.
   if (inputs.front().device().type() == at::kLazy) {
```

Comment:
to double check,
* we would never pass empty `inputs` to this function? 
* can we end up with a mix of devices cuda and lazy?

Reply:
Not sure about empty `inputs`. Even if they are empty, how does that matter?
Currently, we only hack it for broadcast and allReduce. For these two ops, we should only have one element in the inputs. Since this is just a hack, not sure if we really need additional assertions or bailouts.